core/src/commands/publish_datapoint.rs

```diff
@@ -0,0 +1,381 @@
+use std::convert::TryInto;
+
+use derive_more::From;
+use ergo_lib::{
+    chain::ergo_box::box_builder::{ErgoBoxCandidateBuilder, ErgoBoxCandidateBuilderError},
+    ergotree_interpreter::sigma_protocol::prover::ContextExtension,
+    ergotree_ir::chain::{
+        address::Address,
+        ergo_box::{
+            box_value::BoxValue,
+            NonMandatoryRegisterId::{R4, R5, R6},
+        },
+    },
+    wallet::{
+        box_selector::{BoxSelection, BoxSelector, BoxSelectorError, SimpleBoxSelector},
+        tx_builder::{TxBuilder, TxBuilderError},
+    },
+};
+use ergo_node_interface::node_interface::NodeError;
+use thiserror::Error;
+
+use crate::{
+    actions::PublishDataPointAction,
+    box_kind::{OracleBox, PoolBox},
+    oracle_state::{LocalDatapointBoxSource, PoolBoxSource, StageError},
+    wallet::WalletDataSource,
+};
+
+#[derive(Debug, Error, From)]
+pub enum PublishDatapointActionError {
+    #[error("stage error: {0}")]
+    StageError(StageError),
+    #[error("Oracle box has no reward token")]
+    NoRewardToken,
+    #[error("tx builder error: {0}")]
+    TxBuilder(TxBuilderError),
+    #[error("box builder error: {0}")]
+    ErgoBoxCandidateBuilder(ErgoBoxCandidateBuilderError),
+    #[error("node error: {0}")]
+    Node(NodeError),
+    #[error("box selector error: {0}")]
+    BoxSelector(BoxSelectorError),
+}
+
+pub fn build_publish_datapoint_action(
+    pool_box_source: &dyn PoolBoxSource,
+    local_datapoint_box_source: &dyn LocalDatapointBoxSource,
+    wallet: &dyn WalletDataSource,
+    height: u32,
+    change_address: Address,
+    new_datapoint: i64,
+) -> Result<PublishDataPointAction, PublishDatapointActionError> {
+    let in_pool_box = pool_box_source.get_pool_box()?;
+    let in_oracle_box = local_datapoint_box_source.get_local_oracle_datapoint_box()?;
```

Comment:
I believe we should handle a case when oracle just joined and there is no existing data point box. For this case box.value could be `BoxValue::SAFE_USER_MIN` and ergo tree from OracleContract. The public key for R4 should be set in the oracle conf file and the oracle token could be found in the wallet's unspent boxes via box selector (we can assume oracle has it at this point).

Reply:
Thanks for the details, will do.

Reply:
Made the changes. Had to modify `OraclePool` struct. In particular the `local_oracle_datapoint_scan` field is now optional, to allow for the publishing of the first datapoint.